python/tank/bootstrap/configuration_writer.py

```diff
@@ -28,6 +28,50 @@
 log = LogManager.get_logger(__name__)
 
 
+class RenameGuard(object):
+    """
+    Ensures that file that were renamed during a scope are renamed to their
+    original name if an exception was raised during that scope.
+    """
+
+    def __init__(self, undo_on_error):
+        """
+        :param bool undo_on_error: If true, the renames will be undone when an exception
+            is raised. If false, the files won't be renamed when an exception is raised.
+        """
+        self._undo_on_error = undo_on_error
+        self._renames = []
+
+    def __enter__(self):
+        """
+        Returns itself so files can be renamed and tracked.
+        """
+        return self
+
+    def rename(self, source, dest):
+        """
+        Renames a file and keeps track of the rename if it succeeded.
+
+        :param str source: File to rename.
+        :param str dest: New name of the file.
+        """
+        os.rename(source, dest)
+        self._renames.append((source, dest))
```

Comment:
If `os.rename` raises in the previous line, will this line be executed? If not, will it be able to undo? Also, what scenarios would this raise after the file was moved? i.e. perhaps this could be simplified to a `copy` + `delete` if `undo_on_error` was supplied? Sorry, just want to make sure I understand all the scenarios.